### PR TITLE
Provide structured suggestion for type mismatch in loop

### LINKED ISSUE #98982: Detect type mismatch due to loop that might never iterate
Given

```rust
fn foo() -> i32 {
    for i in 0..0 {
        return i;
    }
}
```
the current output is
```
error[E0308]: mismatched types
 --> src/lib.rs:2:5
  |
1 |   fn foo() -> i32 {
  |               --- expected `i32` because of return type
2 | /     for i in 0..0 {
3 | |         return i;
4 | |     }
  | |_____^ expected `i32`, found `()`
```

If a `return` statement is present in the tail expression, and that tail expression is a loop, it should provide more information for newcomers:

```
error[E0308]: mismatched types
 --> src/lib.rs:2:5
  |
1 |   fn foo() -> i32 {
  |               --- expected `i32` because of return type
2 | /     for i in 0..0 {
3 | |         return i;
4 | |     }
  | |_____^ expected `i32`, found `()`
note: the function expects a value to always be returned, but loops might run zero times
 --> src/lib.rs:2:5
  |
2 |     for i in 0..0 {
  |              ^^^^ this might have zero elements to iterate on
3 |         return i;
  |                - if the loop doesn't execute, this value would never get returned
help: return a value for the case when the loop has zero elements to iterate on, and consider changing the return type to account for that possibility
  |
1 ~ fn foo() -> Option<i32> {
...
5 +     None
  |
```

### PR DESCRIPTION
**Pull request by estebank**
We currently provide only a `help` message, this PR introduces the last two structured suggestions instead:

```
error[E0308]: mismatched types
  --> $DIR/issue-98982.rs:2:5
   |
LL |   fn foo() -> i32 {
   |               --- expected `i32` because of return type
LL | /     for i in 0..0 {
LL | |         return i;
LL | |     }
   | |_____^ expected `i32`, found `()`
   |
note: the function expects a value to always be returned, but loops might run zero times
  --> $DIR/issue-98982.rs:2:5
   |
LL |     for i in 0..0 {
   |     ^^^^^^^^^^^^^ this might have zero elements to iterate on
LL |         return i;
   |         -------- if the loop doesn't execute, this value would never get returned
help: return a value for the case when the loop has zero elements to iterate on
   |
LL ~     }
LL ~     /* `i32` value */
   |
help: otherwise consider changing the return type to account for that possibility
   |
LL ~ fn foo() -> Option<i32> {
LL |     for i in 0..0 {
LL ~         return Some(i);
LL ~     }
LL ~     None
   |
```

Fix #98982.

### CODE REVIEW

**Comment by cjgillot** [comment]
Did you forget part of the commit?

**Comment by estebank** [comment]
@cjgillot I apparently had not staged the changes I wanted 😅

File: tests/ui/typeck/issue-100285.stderr [hunk 1]
```diff
@@ -27,7 +27,32 @@ LL |         } else if n < 20 {
...
+LL ~     }
+LL ~     /* `i32` value */
+   |
+help: otherwise consider changing the return type to account for that possibility
+   |
+LL ~ fn foo(n: i32) -> Option<i32> {
+LL |     for i in 0..0 {
+LL |        if n < 0 {
+LL ~         return Some(i);
+LL |         } else if n < 10 {
```
Comment (cjgillot) [inline tests/ui/typeck/issue-100285.stderr:41]: The indentation looks weird.

Comment (cjgillot) [reply tests/ui/typeck/issue-100285.stderr:41]: I just saw, preexisting.

**Pull request merged**